cpp/src/arrow/array/array_nested.cc

```diff
@@ -263,7 +381,7 @@ Result<std::shared_ptr<Array>> FlattenListArray(const ListArrayT& list_array,
   return Concatenate(non_null_fragments, memory_pool);
 }
 
-template <typename ListViewArrayT, bool HasNulls>
+template <typename ListViewArrayT, bool HasNulls, bool IncludeNulls>
```

Comment:
templatizing with an addtional 3rd arg (IncludeNulls) makes the code harder to reason about.